src/Compilers/CSharp/Portable/Binder/Binder_Patterns.cs

```diff
@@ -175,11 +175,261 @@ private BoundExpression BindSwitchExpression(SwitchExpressionSyntax node, Bindin
+            ref bool hasErrors,
+            out bool sawSlice,
+            BindingDiagnosticBag diagnostics)
+        {
+            Debug.Assert(node.IsKind(SyntaxKind.ListPatternClause));
+
+            sawSlice = false;
+            var subpatterns = node.Subpatterns;
+            var builder = ArrayBuilder<BoundPattern>.GetInstance(subpatterns.Count);
+            foreach (SubpatternSyntax subpat in subpatterns)
+            {
+                // Any name would have make this a PropertyPatternClause
+                Debug.Assert(subpat.NameColon is null);
+
+                PatternSyntax pattern = subpat.Pattern;
+                BoundPattern boundPattern;
+                if (pattern is SlicePatternSyntax slice)
+                {
+                    boundPattern = BindSlicePattern(slice, inputType, permitDesignations, ref hasErrors, misplaced: sawSlice, diagnostics);
+                    sawSlice = true;
+                }
+                else
+                {
+                    boundPattern = BindPattern(pattern, elementType, ExternalScope, permitDesignations, hasErrors, diagnostics);
+                }
+
+                builder.Add(boundPattern);
+            }
+
+            return builder.ToImmutableAndFree();
+        }
+
+        private BoundPattern BindLengthPatternClause(
+            LengthPatternClauseSyntax node, TypeSymbol inputType,
+            bool permitDesignations, out PropertySymbol? lengthProperty, ref bool hasErrors, BindingDiagnosticBag diagnostics)
+        {
+            LookupLengthOrCountProperty(node, inputType, out lengthProperty, ref hasErrors, diagnostics);
+            return BindPattern(node.Pattern, Compilation.GetSpecialType(SpecialType.System_Int32), ExternalScope, permitDesignations, hasErrors, diagnostics);
+        }
+
+        private void LookupLengthOrCountProperty(
+            SyntaxNode node, TypeSymbol inputType, out PropertySymbol? lengthProperty,
+            ref bool hasErrors, BindingDiagnosticBag diagnostics)
+        {
+            Debug.Assert(node.IsKind(SyntaxKind.ListPatternClause) || node.IsKind(SyntaxKind.LengthPatternClause));
+
+            using var _ = LookupResult.GetInstance(out var lookupResult);
+            var useSiteInfo = CompoundUseSiteInfo<AssemblySymbol>.Discarded;
+            if (!TryLookupLengthOrCount(inputType, lookupResult, out lengthProperty, ref useSiteInfo))
+            {
+                if (!inputType.IsErrorType() && !hasErrors)
+                {
+                    var errorCode = node.IsKind(SyntaxKind.ListPatternClause)
+                        ? ErrorCode.ERR_UnsupportedTypeForListPattern
+                        : ErrorCode.ERR_UnsupportedTypeForLengthPattern;
+                    Error(diagnostics, errorCode, node, inputType.ToDisplayString());
+                }
+                hasErrors = true;
+            }
+        }
+
+        private BoundListPatternClause BindListPatternClause(
+            PropertyPatternClauseSyntax node,
```

Comment:
Could we also just pass in the `ListPatternCauseSyntax` here?

Reply:
This is already doing that. There's no such node as ListPatternCauseSyntax.